### PR TITLE
Add option to normalize NS input in double precision

### PR DESCRIPTION
**Pull request by skyw**
"if" clause on calculated norm would trigger device to host sync, trying to do it on device will take multiple path, in which case none of them are better than just use double. A custom kernel can do it but is out of scope.

No zero division guard for fp64 path as we don't imagine a case that LLM training is done in fp64. Square of fp32 value won't underflow fp64.

### CODE REVIEW

File: emerging_optimizers/orthogonalized_optimizers/muon_utils.py [hunk 1]
```diff
@@ -130,10 +131,18 @@ def get_coefficient_iterator(
...
-    return x / torch.sqrt(x_sq_sum).clamp_min(eps)
+    norm = torch.sqrt(x_sq_sum)
+    if norm < eps:
+        shift = torch.ceil(torch.where(norm > 0, torch.log2(eps / norm), math.log2(1 / eps)))
+        x = x * torch.exp2(shift)
+        x_sq_sum = (x * x).sum()
+        torch.distributed.all_reduce(x_sq_sum, op=torch.distributed.ReduceOp.SUM, group=group)
+        norm = torch.sqrt(x_sq_sum)
+        assert norm > 0  # Fail if it still underflows
+    return x / norm
```
[the remark this answers is not in the document]
Comment (skyw) [reply emerging_optimizers/orthogonalized_optimizers/muon_utils.py]: @greptile assert internal assumption is not an issue, should be more concerned about the device to host sync

[the remark this answers is not in the document]
Comment (skyw) [reply emerging_optimizers/orthogonalized_optimizers/muon_utils.py]: @greptileai remember any syncs are high priority, regardless of under which context.

File: emerging_optimizers/orthogonalized_optimizers/muon_utils.py [hunk 2]
```diff
@@ -129,11 +129,21 @@ def get_coefficient_iterator(
...
+
+    When ``normalize_in_double`` is set, the squared sum is accumulated in float64 so that tiny
+    entries do not underflow to zero when squared in float32.
+    """
+    x_sq = x.double() if normalize_in_double else x
+    x_sq_sum = (x_sq * x_sq).sum()
     torch.distributed.all_reduce(x_sq_sum, op=torch.distributed.ReduceOp.SUM, group=group)
-    return x / torch.sqrt(x_sq_sum).clamp_min(eps)
+    norm = torch.sqrt(x_sq_sum).to(x.dtype)
+    if not normalize_in_double:
```
Comment (FDecaYed) [inline emerging_optimizers/orthogonalized_optimizers/muon_utils.py:144]: I feel normalizing in double not prevent tiny entries to not become 0, thus preventing some cases where norm is too small(or 0). But in general it should not cause eps clamp to be skipped entirely?
- casting down of norm can still produce 0
- input can be all 0 to begin with
though it could also be me misunderstanding what this eps is guarding in the first place

Comment (skyw) [reply emerging_optimizers/orthogonalized_optimizers/muon_utils.py:144]: In theory, it does not.
But 0 in normalization is usually (if not always) caused by sum(x*x) underflow but sqrt(sum(x*x)) should still be representable. Add an eps is a very naive way to move on. Because we'll only ever train in <=fp32 precision, x*x won't underflow double precision, so it won't be 0.

In the extreme case where entire input is literary 0, return inf is arguably the right behavior.

**Review by FDecaYed — APPROVED** [review]
